File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/Emoji.java
package ja.burhanrashid52.photoeditor;

import android.graphics.Typeface;
import android.view.Gravity;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

/**
 * Created by Burhanuddin Rashid on 14/05/21.
 *
 * @author <https://github.com/burhanrashid52>
 */
class Emoji extends Graphic {

    private final MultiTouchListener mMultiTouchListener;
    private final Typeface mDefaultEmojiTypeface;
    private final OnPhotoEditorListener mOnPhotoEditorListener;
    private final RelativeLayout mCanvasView;
    private final PhotoEditorViewState mViewState;
    private TextView txtEmoji;

    public Emoji(RelativeLayout canvasView,
                 MultiTouchListener multiTouchListener,
                 PhotoEditorView photoEditorView,
                 PhotoEditorViewState viewState,
                 OnPhotoEditorListener onPhotoEditorListener,
                 GraphicManager graphicManager,
                 Typeface defaultEmojiTypeface
    ) {
        super(photoEditorView.getContext(), graphicManager);
        mCanvasView = canvasView;
        mOnPhotoEditorListener = onPhotoEditorListener;
        mViewState = viewState;
        mMultiTouchListener = multiTouchListener;
        mDefaultEmojiTypeface = defaultEmojiTypeface;
        setupGesture();
    }

    void buildView(Typeface emojiTypeface, String emojiName) {
        if (emojiTypeface != null) {
            txtEmoji.setTypeface(emojiTypeface);
        }
        txtEmoji.setTextSize(56);
        txtEmoji.setText(emojiName);
    }

    private void setupGesture() {
        MultiTouchListener.OnGestureControl onGestureControl = buildGestureController(
                mCanvasView,
                mViewState,
                mOnPhotoEditorListener
        );
        mMultiTouchListener.setOnGestureControl(onGestureControl);
        View rootView = getRootView();
        rootView.setOnTouchListener(mMultiTouchListener);
    }

    @Override
    ViewType getViewType() {
        return ViewType.EMOJI;
    }

    @Override
    int getLayoutId() {
        return R.layout.view_photo_editor_text;
    }

    @Override
    void setupView(View rootView) {
        txtEmoji = rootView.findViewById(R.id.tvPhotoEditorText);
        if (txtEmoji != null) {
            if (mDefaultEmojiTypeface != null) {
                txtEmoji.setTypeface(mDefaultEmojiTypeface);
            }
            txtEmoji.setGravity(Gravity.CENTER);
            txtEmoji.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        }
    }
}

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/Emoji.kt
package ja.burhanrashid52.photoeditor

import android.graphics.Typeface
import android.view.Gravity
import android.view.View
import android.widget.TextView

/**
 * Created by Burhanuddin Rashid on 14/05/21.
 *
 * @author <https:></https:>//github.com/burhanrashid52>
 */
internal class Emoji(
    private val mPhotoEditorView: PhotoEditorView,
    private val mMultiTouchListener: MultiTouchListener,
    private val mViewState: PhotoEditorViewState,
    graphicManager: GraphicManager?,
    private val mDefaultEmojiTypeface: Typeface?
) : Graphic(
    context = mPhotoEditorView.context,
    graphicManager = graphicManager,
    viewType = ViewType.EMOJI,
    layoutId = R.layout.view_photo_editor_text
) {
    private var txtEmoji: TextView? = null
    fun buildView(emojiTypeface: Typeface?, emojiName: String?) {
        txtEmoji?.apply {
            if (emojiTypeface != null) {
                typeface = emojiTypeface
            }
            textSize = 56f
            text = emojiName
        }
    }

    private fun setupGesture() {
        val onGestureControl = buildGestureController(mPhotoEditorView, mViewState)
        mMultiTouchListener.setOnGestureControl(onGestureControl)
        val rootView = rootView
        rootView.setOnTouchListener(mMultiTouchListener)
    }

    override fun setupView(rootView: View) {
        txtEmoji = rootView.findViewById(R.id.tvPhotoEditorText)
        txtEmoji?.run {
            if (mDefaultEmojiTypeface != null) {
                typeface = mDefaultEmojiTypeface
            }
            gravity = Gravity.CENTER
            setLayerType(View.LAYER_TYPE_SOFTWARE, null)
        }
    }

    init {
        setupGesture()
    }
}